File: site-src/examples/algorithm-deepening/pattern-lab-v04/monotonic_structures_trace.cpp
#include <deque>
#include <iostream>
#include <optional>
#include <stdexcept>
#include <vector>

struct GreaterResult {
  std::vector<std::optional<long long>> answers;
  std::size_t resolved;
  std::size_t unresolved;
};

GreaterResult next_strictly_greater(const std::vector<long long>& values) {
  std::vector<std::optional<long long>> answers(values.size());
  std::vector<std::size_t> stack;
  std::size_t resolved = 0;
  for (std::size_t index = 0; index < values.size(); ++index) {
    while (!stack.empty() && values[stack.back()] < values[index]) {
      answers[stack.back()] = values[index];
      stack.pop_back();
      ++resolved;
    }
    stack.push_back(index);
  }
  return {answers, resolved, stack.size()};
}

struct MaximumResult {
  std::vector<long long> maxima;
  std::size_t back_pruned;
  std::size_t expired;
};

MaximumResult sliding_maximum(const std::vector<long long>& values,
                              std::size_t width) {
  if (width == 0) throw std::invalid_argument("width must be positive");
  if (width > values.size()) return {};
  std::deque<std::size_t> candidates;
  MaximumResult result{};
  for (std::size_t index = 0; index < values.size(); ++index) {
    while (!candidates.empty() && candidates.front() + width <= index) {
      candidates.pop_front();
      ++result.expired;
    }
    while (!candidates.empty() &&
           values[candidates.back()] <= values[index]) {
      candidates.pop_back();
      ++result.back_pruned;
    }
    candidates.push_back(index);
    if (index + 1 >= width) {
      result.maxima.push_back(values[candidates.front()]);
    }
  }
  return result;
}

int main() {
  const std::vector<long long> values{2, 1, 2, 4, 3};
  const auto greater = next_strictly_greater(values);
  const auto maximum = sliding_maximum(values, 3);
  std::cout << "values=2,1,2,4,3\nnext_strictly_greater=";
  for (std::size_t index = 0; index < greater.answers.size(); ++index) {
    if (index > 0) std::cout << ',';
    if (greater.answers[index].has_value()) std::cout << *greater.answers[index];
    else std::cout << "none";
  }
  std::cout << "\nstack_resolved=" << greater.resolved
            << " unresolved=" << greater.unresolved << '\n';
  std::cout << "window_width=3\nwindow_maxima=";
  for (std::size_t index = 0; index < maximum.maxima.size(); ++index) {
    if (index > 0) std::cout << ',';
    std::cout << maximum.maxima[index];
  }
  std::cout << "\ndeque_back_pruned=" << maximum.back_pruned
            << " expired=" << maximum.expired << '\n';
  std::cout << "invariant=dominated-candidates-never-return\n";
}

File: site-src/examples/algorithm-deepening/pattern-lab-v04/monotonic_structures_trace.py
from __future__ import annotations

from collections import deque
from dataclasses import dataclass
from typing import Sequence


@dataclass(frozen=True)
class NextGreaterResult:
    answers: tuple[int | None, ...]
    resolved: int
    unresolved: int


def next_strictly_greater(values: Sequence[int]) -> NextGreaterResult:
    answers: list[int | None] = [None] * len(values)
    stack: list[int] = []
    resolved = 0
    for index, value in enumerate(values):
        while stack and values[stack[-1]] < value:
            answers[stack.pop()] = value
            resolved += 1
        stack.append(index)
    return NextGreaterResult(tuple(answers), resolved, len(stack))


@dataclass(frozen=True)
class WindowMaximumResult:
    maxima: tuple[int, ...]
    back_pruned: int
    expired: int


def sliding_maximum(values: Sequence[int], width: int) -> WindowMaximumResult:
    if width <= 0:
        raise ValueError("width must be positive")
    if width > len(values):
        return WindowMaximumResult((), 0, 0)
    candidates: deque[int] = deque()
    maxima: list[int] = []
    back_pruned = 0
    expired = 0
    for index, value in enumerate(values):
        while candidates and candidates[0] <= index - width:
            candidates.popleft()
            expired += 1
        while candidates and values[candidates[-1]] <= value:
            candidates.pop()
            back_pruned += 1
        candidates.append(index)
        if index + 1 >= width:
            maxima.append(values[candidates[0]])
    return WindowMaximumResult(tuple(maxima), back_pruned, expired)


def fixed_report() -> str:
    values = [2, 1, 2, 4, 3]
    greater = next_strictly_greater(values)
    maximum = sliding_maximum(values, 3)
    answers = ",".join("none" if item is None else str(item) for item in greater.answers)
    return "\n".join(
        [
            "values=2,1,2,4,3",
            f"next_strictly_greater={answers}",
            f"stack_resolved={greater.resolved} unresolved={greater.unresolved}",
            "window_width=3",
            f"window_maxima={','.join(map(str, maximum.maxima))}",
            f"deque_back_pruned={maximum.back_pruned} expired={maximum.expired}",
            "invariant=dominated-candidates-never-return",
        ]
    )


if __name__ == "__main__":
    print(fixed_report())
